File: 2020/25/solution.hs
d = 20201227

transform :: Int -> Int -> Int
transform current subject = (current * subject) `rem` d

transforms :: Int -> [Int]
transforms subject = iterate (transform subject) 1

loopSize :: Int -> Int -> Int
loopSize publicKey subject =
    snd n
  where
    n = last $ takeWhile (\(x, y) -> x /= publicKey) $ enumerate $ transforms subject

enumerate :: [a] -> [(a, Int)]
enumerate list = zip list [1..]

solution :: Int -> Int -> Int
solution publicKey1 publicKey2 =
    last $ take (loop1size + 1) $ transforms publicKey2
  where
    loop1size = loopSize publicKey1 7

main = do
  putStr $ "Part 1 answer is: "
  putStr $ show $ solution 8987316 14681524
  putStr "\n"

File: 2020/25/solution.rb
DIV = 20201227

def transform(current_value, subject_value)
  current_value *= subject_value

  current_value %= DIV
end

def loop_size(public_key, subject)
  current = 1

  i = 0

  loop do
    current = transform(current, subject)
    i+= 1

    break if current == public_key
  end

  i
end

def apply_transform(subject_value, loop_size)
  current = 1

  loop_size.times do
    current = transform(current, subject_value)
  end

  current
end

public_key1 = 8987316
public_key2 = 14681524

loop_size1 = loop_size(public_key1, 7)
loop_size2 = loop_size(public_key2, 7)

puts apply_transform(public_key2, loop_size1)
